### ISSUE TITLE
Add Caching layer for OLAP queries

### ISSUE THREAD

**Issue by etiennedi** [creativesoftwarefdn/weaviate#720]
## Todos

* [x] Upgrade Janus to 0.3.x
* [x] Build Java analytics service
  * [x] setup CI/dev environment
  * [x] take query and request id in request body
  * [x] store different events in etcd
    * [x] starting
    * [x] finished
    * [x] failed
  * [x] add acceptance test
  *  ~~add CI pipeline to push to dockerhub~~ (outscoped to https://github.com/semi-technologies/janus-spark-analytics/issues/3)
* [x] modify weaviate
  * [x] allow config as yaml (inscoped from #777) 
  * [x] allow spark to be configured as part of janusgraph connector
  * [x] introduce flag to decide whether analytics engine should be used per query
    * [x] `useAnalyticsEngine` (see comments below)
    * [x] `forceRecalculate` (see comments below)
    * [x] in `Local.GetMeta`
    * [x] in `Local.Aggregate`
    * [x] validate if combination makes sense
    * [x] pass info to connectors
    * [x] calculate hash 
      * the cache settings cannot affect the hash otherwise `forceRecalculate`d results can never be retrieved from the cache again
      * [x] for GetMeta
      * [x] for Aggregate
  * [x] if yes, check for cache and start or retrieve job results in connector
  * [x] rebuild all queries so they work with Spark
    * ~~fix issues with counting cross-refs (Aggregate)~~ - outscoped to #779 
  * [x] add option to invalidate cache
  * ~~add acceptance test (will require heavy fakes)~~
    * I decided against a new acceptance test since most of the changes in weaviate are the updated query structure which is already covered by existing acceptance tests. Additionally we have added some new logic that decides whether the cache should be used or a new job triggered. This is fully covered with unit tests. Finally, the spark analytics journey is fully tested in https://github.com/semi-technologies/janus-spark-analytics

## Original requirements
* OLAP queries take a long time (minutes to hours) to complete
* To keep a good UX we want to be able to send of an OLAP query and let the user come back to check if it's finished later. To do so they can send the exact same query again. If it's in the cache it will be served
* There needs to be a "force recreate" option to force start a new analytical job even if the cache is filled
* We are currently not aware of how to deal with the fact that from the gremlin console OLAP queries (to Spark) are blocking - this is an outstanding help request to external consultants.

**Comment by etiennedi** [comment]
should be tackled after #719

**Comment by etiennedi** [comment]
Decision log from Slack conversations between @bobvanluijt and me:

## Question: Do we need a flag to turn caching off?
### When might users not want caching?
* If they have an inmemory connector for "small, but fast" graphs
* If they know their graph is small enough that it can calculate OLAP on the fly regardless of the connector type

### Decision:
* We will introduce a flag to turn "caching for OLAP" on or off somewhere in the db config 
* This flag exists regardless of the connector, but the implications might differ from connector to connector
* If the flag is turned off no cache should be used, this also means no "cache invalidation" option should be available
* If the flag is turned on, each connector can decide what to do
  * An inmemory connector could fail stating that the in-mem connector should not have a cache
  * The janus connector can use this flag to decide whether to use Spark for OLAP or try and do it without spark (and without a cache)

**Comment by etiennedi** [comment]
## User experience impact of long-running analysis job and caching

This example assumes an analysis job takes 30min to complete. (It deals with both this issue, as well as #719)

### What changes for the user?
Prior to this, we ignored the fact that analysis jobs take considerable time on large graphs. We assumed that results would be real-time.

This aims to keep the UX as nice as possible. For the user the changes should be minimal. Therefore the proposal would be:

#### Iteration 1
Nothing changes about the way the user submits a `GetMeta` or `Aggregate` query. However, they can get one of three possible results:

1. State: Error - Message: "No cached analysis present. An analysis job has been triggered, please check back later" -> in the background we trigger a spark job
2. State: Error - Message: "An analysis job is currently running, but has not completed yet, please check back later" -> this would appear if the user just immediately re-sends their request
3. State: OK -> Present the results from cache.

#### Iteration 2
Assuming the caching period is 24h. The user is aware that they made considerable changes to the dataset after one hour and don't want to wait another 23h to be able to submit a new analysis:

Proposal: The user can include a `forceRecalculate: true` (<-- name is not ideal, input appreciated) option in the request. For example:`Local { GetMeta(forceRecreate: true) { ... } }` If this flag is set, we don't check for a cached result, invalidate the cache and start a new spark job.

@laura-ham Feedback for both Iteration 1 and Iteration 2 highly appreciated. Thanks.

CC @bobvanluijt

**Comment by bobvanluijt** [comment]
Thanks @etiennedi. @laura-ham, what would be the appropriate place to add this UX to the docs?

**Comment by laura-ham** [comment]
@etiennedi 
**Iteration 1**
This is a good way of giving the user feedback of what is happening. Is it possible to give a time estimation of when the caching should be done? Then the user knows what he can expect from the machine. 

**Iteration 2**
I think either `forceRecalculate` or `forceRecreate` (prefer the former) (or `forceReload`) are appropriate names for the functions, because this is exactly what happens. 

@bobvanluijt I think this concept should be introduced at the starting page of Graphql (https://github.com/creativesoftwarefdn/weaviate/blob/develop/docs/en/use/graphql_introduction.md), and then we can mention the arguments as well at the specific `Local` and `Network` pages, referring to the `Introduction` page for details?

**Comment by bobvanluijt** [comment]
Check @laura-ham, feel free to add it. But please do this in the `feature/docs` branch

**Comment by etiennedi** [comment]
@laura-ham 
I think I completely missed your response. Sorry about that. I think we've already talked about the time estimation on Slack, so for completeness sake (and other readers) here's a self-quote from the slack conversation:

>etienne   [3 days ago]
As for prediction of time, if the query has ever run before we could tell the user „last time it took x minutes/hours“. But if it’s the first time we have no way of knowing. Or we could build an average over all queries and give the user that as an estimate.
>
>etienne   [3 days ago]
Or both :)

In the meantime here's a small proposal for the api changes and some configuration management ideas, that I wanted to run by you. I'm thinking about adding these two new fields:

### New Fields
```graphql
{
  Local {
    GetMeta {
      Things {
        City(
          forceRecalculate: true,
          useAnalyticsEngine: true,
        ) {
          population {
            sum
          }
        }
      }
    }
  }
}
```

* `forceRecalculate` as outline above
* `useAnalyticsEngine` would tell weaviate whether to use spark or not. The motivation behind being able to turn this off (as outlined in the previous update video) would be that if we have already narrowed down the result set with a filter to let's say less than 1000 results, we don't need to use Spark. The benefit to the user would be that results are instant whereas they could take minutes or hours with Spark.

### Defaults
I think for `forceRecalculate` it makes sense to default to `false`. I would assume that the standard case is that we want to use a cache and the special case is that we want to invalidate the cache. Agreed?

For `useAnalyticsEngine` it's not so easy though. This really depends on the users behavior. If they usually use `where` filters they might only rarely need Spark. If they usually try to analyze the entire data set they will want to use Spark most of the time. So my best suggestion would be to make the default configurable in the application wide config. Opinions?

### Required fields & Validation
If we have reasonable defaults we don't have to make the fields required. However, we could also say there are no defaults and instead the user will always have to specify the desired behavior. (<- Not the biggest fan of this, as it will it look harder to use, since so many fields have to be filled).

There is one case where validation must fail though: 
```json
       forceRecalculate: true,
       useAnalyticsEngine: false,
```
This is not a valid combination. As there is no cached version that could be used (or invalidated) if Spark is not meant to be used. So I think this is an the only combination where we should error.

Happy to hear your thoughts and opinions.

**Comment by laura-ham** [comment]
Hi @etiennedi,

> I'm thinking about adding these two new fields.
Sounds good, also for the defaults. One thing is that if we allow or ask the user to set the `useAnalyticsEngine` in the configuration, the user has to think about this beforehand.

So it sounds like that these arguments are dependent on 
* _the config_, which assumes that the two systems are implemented correctly before the user starts querying (otherwise there is just a simple error I suppose, but then the user must be told where to edit). And then the user should of course be able to override this setting in the query itself. 
* _Use of Spark_. This makes the GraphQL design dependent on the config (just like the setting of the argument in the config). I have not experimented with a GQL design dependent on the config, but I assume that this is not going to be a problem, also not UX-wise.

**Comment by etiennedi** [comment]
> * _the config_, which assumes that the two systems are implemented correctly before the user starts querying (otherwise there is just a simple error I suppose, but then the user must be told where to edit). And then the user should of course be able to override this setting in the query itself.

I think the important part is that GraphQL doesn't depend on Janus-specific config. Otherwise it will become a mess for other connectors and we lose the full plugability. So I think in the config we'll need two sections, roughly like so. I'm also adding some comments that we might use in the default configuration if we go this route:

```yaml
name: myConfigWithJanusgraphAndSpark
analyticsEngine:
  # enabled configures whether an analytics engine should be used. 
  # Setting this to true leads to the options "useAnalyticsEngine" and
  # "forceRecalculate" to become available in the GraphQL GetMeta->Kind->Class
  # and Aggregate->Kind->Class.
  #
  # Important: If enabled is set to true, you must also configure an
  # analytics engine in your database connector. If the chosen connector
  # does not support an external analytics engine, enabled must be set
  # to false.
  enabled: true  
  
  # defaultUseAnalyticsEngine configures what the "useAnalyticsEngine"
  # in the GraphQL API will default to when not set.
  defaultUseAnalyticsEngine: true 
database:
  name: janusgraph
  database_config:  # this object is evaluated by the connector, the rest have the app has no access to it
    url: http://janus:8182
    analyticsEngine:
      # enabled enables an analytics engine for this particular connector
      # Important: If enabled is set to true, the overall configuration property
      # analyticsEngine.enabled must also be set to true. 
      enabled: true

      # type configures which analytics engine is used with this connector
      # currently the janusgraph connector supports only 'spark-analytics-api'
      type: spark-analytics-api

      # url to discover the specific analytics backend
      url: http://analytics-api:8080
```

⬆️ @bobvanluijt @laura-ham Feedback on the above appreciated.

> * _Use of Spark_. This makes the GraphQL design dependent on the config (just like the setting of the argument in the config). I have not experimented with a GQL design dependent on the config, but I assume that this is not going to be a problem, also not UX-wise.

I think this is quite similar to how the GraphQL schema is already dynamically built based on the class schema. I also don't see any issues there, neither from a usage nor from an implementation perspective.

**Comment by laura-ham** [comment]
Looks good!

**Comment by etiennedi** [comment]
Minor change to what I suggested above: The remainder of the config uses `snake_case` rather than `camel_case`, so the new fields should be consistent with that as well.

**Comment by etiennedi** [comment]
@laura-ham @bobvanluijt 
Just realized that  - since we also have `Network.GetMeta` and `Network.Aggregate` - technically we'd need to add those fields there as well. However, I'd suggest that we view Spark over Network as an edge case for now. Since we only proxy, it can still work, however the UX would be slightly worse than on a local query.

For the local query we have the options outlined as in the previous 5 posts. We can't really provide them in the same fashion for all network peers without extending the way the p2p network works at the moment. Otherwise we won't know which peers support external analytical engines. (Adding this to the network is possible of course, but noticable effort and more than is necessary to reach this goal in my opinion). 

Therefore my suggestions would be:

1.) For now, let's just disable queries with the analytical engine for `Network` queries
2.) Let's think about if we ever want to allow them. Since they can be very expensive computationally, it's an interesting question whether any peer should be able to trigger such workloads for any other peer
3.) Only if we decide that we actually want to support it, let's worry about the how.

Please let me know if you disagree with this approach.
